map.go

```diff
@@ -420,34 +420,41 @@ func (spec *MapSpec) createMap(inner *sys.FD, opts MapOptions) (_ *Map, err erro
 		attr.MapName = sys.NewObjName(spec.Name)
 	}
 
-	btfDisabled := false
-	if spec.Type.supportsBTF() && (spec.Key != nil || spec.Value != nil) {
+	if spec.Key != nil || spec.Value != nil {
 		handle, keyTypeID, valueTypeID, err := btf.MarshalMapKV(spec.Key, spec.Value)
-		btfDisabled = errors.Is(err, btf.ErrNotSupported)
-		if err != nil && !btfDisabled {
+		if err != nil && !errors.Is(err, btf.ErrNotSupported) {
 			return nil, fmt.Errorf("load BTF: %w", err)
 		}
+
 		if handle != nil {
 			defer handle.Close()
 
+			// Use BTF k/v during map creation.
 			attr.BtfFd = uint32(handle.FD())
 			attr.BtfKeyTypeId = uint32(keyTypeID)
 			attr.BtfValueTypeId = uint32(valueTypeID)
 		}
 	}
 
 	fd, err := sys.MapCreate(&attr)
+	// Some map types don't support BTF k/v in earlier kernel versions.
+	// Remove BTF metadata and retry map creation.
+	if (errors.Is(err, sys.ENOTSUPP) || errors.Is(err, unix.EINVAL)) && attr.BtfFd != 0 {
+		attr.BtfFd, attr.BtfKeyTypeId, attr.BtfValueTypeId = 0, 0, 0
+		fd, err = sys.MapCreate(&attr)
+	}
+
 	if err != nil {
 		if errors.Is(err, unix.EPERM) {
 			return nil, fmt.Errorf("map create: %w (MEMLOCK may be too low, consider rlimit.RemoveMemlock)", err)
 		}
 		if errors.Is(err, unix.EINVAL) && attr.MaxEntries == 0 {
 			return nil, fmt.Errorf("map create: %w (MaxEntries may be incorrectly set to zero)", err)
 		}
-		if btfDisabled {
-			return nil, fmt.Errorf("map create: %w (kernel without BTF support)", err)
+		if attr.BtfFd == 0 {
+			return nil, fmt.Errorf("map create without BTF k/v: %w", err)
 		}
-		return nil, fmt.Errorf("map create: %w", err)
```

Comment:
Nit: I'd keep this error message the same, the whole with / without BTF thingie is confusing to users. The happy case should have BTF (?) so if it'd be nice to keep that simple.

Maybe we should get rid of the "with / without BTF" altogether.

Reply:
> the whole with / without BTF thingie is confusing to users

Not sure if I agree. If the error message is confusing, we should improve it, but IMO having the context is helpful. I've seen this pop up in bug reports on Slack/GH and it makes it much easier to pinpoint which map create actually failed. If we'd have `map create: %w` in multiple places, we need to start adding printf's to figure out which one actually failed. This is especially true in `newProgramWithOptions()`, which has significant complexity around this part.

Reply:
> The happy case should have BTF (?) so if it'd be nice to keep that simple.

Ah, missed that point. Agreed, will do.